File: webapp/go/payment_gateway.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"time"
)

var paymentURL = "http://localhost:12345"

var erroredUpstream = errors.New("errored upstream")

type paymentGatewayPostPaymentRequest struct {
	Token  string `json:"token"`
	Amount int    `json:"amount"`
}

func requestPaymentGatewayPostPayment(param *paymentGatewayPostPaymentRequest) error {
	b, err := json.Marshal(param)
	if err != nil {
		return err
	}

	// 失敗したらとりあえずリトライ
	// FIXME: 社内決済マイクロサービスのインフラに異常が発生していて、同時にたくさんリクエストすると変なことになる可能性あり
	retry := 0
	for {
		err := func() error {
			req, err := http.NewRequest(http.MethodPost, paymentURL+"/payment", bytes.NewBuffer(b))
			if err != nil {
				return err
			}
			req.Header.Set("Content-Type", "application/json")

			res, err := http.DefaultClient.Do(req)
			if err != nil {
				return err
			}
			defer res.Body.Close()

			if res.StatusCode != http.StatusNoContent {
				return fmt.Errorf("unexpected status code (%d): %w", res.StatusCode, erroredUpstream)
			}
			return nil
		}()
		if err != nil {
			if retry < 5 {
				retry++
				time.Sleep(100 * time.Millisecond)
				continue
			} else {
				return err
			}
		}
		break
	}

	return nil
}

File: webapp/python/src/isuride/app/payment_gateway.py
import time
from collections.abc import Callable
from http import HTTPStatus

import urllib3
from pydantic import BaseModel

from isuride.app.models import Ride

ERRORED_UPSTREAM = "errored upstream"


class PaymentGatewayPostPaymentRequest(BaseModel):
    amount: int


class PaymentGatewayGetPaymentsResponseOne(BaseModel):
    amount: int
    status: str


def request_payment_gateway_post_payment(
    payment_gateway_url: str,
    token: str,
    param: PaymentGatewayPostPaymentRequest,
    retrieve_rides_order_by_created_at_asc: Callable[[], list[Ride]],
) -> None:
    # 失敗したらとりあえずリトライ
    # FIXME: 社内決済マイクロサービスのインフラに異常が発生していて、同時にたくさんリクエストすると変なことになる可能性あり
    retry = 0
    while True:
        res = urllib3.request(
            "POST",
            payment_gateway_url + "/payments",
            json=param.model_dump(),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {token}",
            },
        )

        if res.status != HTTPStatus.NO_CONTENT:
            # エラーが返ってきても成功している場合があるので、社内決済マイクロサービスに問い合わせ
            get_res = urllib3.request(
                "GET",
                payment_gateway_url + "/payments",
                headers={
                    "Authorization": f"Bearer {token}",
                },
            )

            # GET /payments は障害と関係なく200が返るので、200以外は回復不能なエラーとする
            if get_res.status != HTTPStatus.OK:
                raise RuntimeError(
                    f"[GET /payments] unexpected status code ({get_res.status})"
                )
            payments = [
                PaymentGatewayGetPaymentsResponseOne(**item) for item in get_res.json()
            ]

            rides = retrieve_rides_order_by_created_at_asc()

            if len(rides) != len(payments):
                raise RuntimeError(
                    f"unexpected number of payments: {len(rides)}  != {len(payments)}. {ERRORED_UPSTREAM}",
                )

        if retry < 5:
            retry += 1
            time.sleep(0.1)
            continue

        break
